### ISSUE TITLE
ActiveMove is problematic

### ISSUE THREAD

**Issue by scheibo** [Zarel/Pokemon-Showdown#5415]
#5389 made `ActiveMove` a class, which brought some concerns:

> Hm I don't think I like the idea of `ActiveMove` being its own class. Especially since it doesn't actually extend `Move`, it actually makes move properties non-readonly.

_Originally posted by @Zarel in https://github.com/Zarel/Pokemon-Showdown/diffs_

and importantly also [caused some crashes](https://github.com/Zarel/Pokemon-Showdown/pull/5389/files#r272753761). It also presents some challenges for serialization in #5270, as I need to include a `_class_` tag (WIP: 'feeling cute, might delete later'):

```json
  "lastMove": {
    "_class_": "ActiveMove",
    "move": "[Move:willowisp]",
    "isViable": true,
    "zMoveBoost": {
      "atk": 1
    },
    "contestType": "Beautiful",
    "hit": 1,
    "moveHitData": {
      "crit": {},
      "typeMod": {},
      "zBrokeProtect": {}
    },
    "totalDamage": 0
  }
```

However, I'm not sure the previous approach (before it was made it a `class`) was any better, as we'd often simply cast arbitrary objects to `/** @type {ActiveMove} */`, and simply cloning + adding crap to a `Data.Move` is [not ideal for performance, as thanks to the monomorphism optimizations](https://mrale.ph/blog/2015/01/11/whats-up-with-monomorphism.html) we would ideally be assigning everything on creation in the literal/`constructor`. From a serialization point of view, detecting an `ActiveMove` without the constructor becomes more difficult (looking for a `hit` field or similar is somewhat error prone, so I'd probably end up not being able to collapse the attributes cloned from `Move`).

- Serialization concerns would have us create an object that looks like:

```js
interface ActiveMove {
  move: Data.Move;
  activeCrap: ...
}
```

- Performance concerns would lead us to not modifying fields of the existing `Move` object, which the `ActiveMove`-as-a-`class` and the strawman interface above would support.
- From conversations with @Zarel it seems like he'd prefer the simplicity of reverting back to the old `ActiveMove` format and simply added a `getHitData` to `Pokemon`.

To be clear, I think 'simplicity' and 'readability' trump 'serialization/performance concerns', but even with the old approach we played [fast](https://github.com/Zarel/Pokemon-Showdown/blob/master/data/mods/gen1/moves.js#L109) and [loose](https://github.com/Zarel/Pokemon-Showdown/blob/master/data/mods/gen4/moves.js#L434) with casts and I don't think the `ActiveMove` concept was optimal.

**Comment by Zarel** [comment]
The previous approach is at least not actively buggy. :/

**Comment by scheibo** [comment]
Yeah, I should mention that fixing the crashes should take precedence over trying to decide what the ideal long term solution should be.
